Add Quickstart Next.js

alex-fusionauth opened the pull request:
New Repo https://github.com/FusionAuth/fusionauth-quickstart-javascript-nextjs-web

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205273462533554
  - https://app.asana.com/0/0/1205297991112218
  - https://app.asana.com/0/0/1205390208575999
  - https://app.asana.com/0/0/1205114579550920

alex-fusionauth commented:
Removed #2347 in favor of being on local branch instead of fork.

mark-robustelli commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,283 @@
+<RemoteCode url={frontmatter.coderoot + "/complete-application/src/app/account/page.tsx"}
+  lang="ts" />
+
+### Run the App!
```
```suggestion
### Run The Application
```

mark-robustelli commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,283 @@
+* [Docker](https://www.docker.com): The quickest way to stand up FusionAuth. (There are [other ways](/docs/v1/tech/installation-guide/)).
+* [git](https://git-scm.com/downloads/): Not required but recommended if you want to track your changes.
+
+## General architecture
```
```suggestion
## General Architecture
```

mark-robustelli commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,283 @@
+## Getting Started
+In this section, you'll get FusionAuth up and running, and configured with the ChangeBank application.
+
+### Clone the Code
```
```suggestion
### Clone The Code
```

mark-robustelli commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
````diff
@@ -0,0 +1,283 @@
+cd fusionauth-quickstart-javascript-nextjs-web
+```
+
+### Run FusionAuth via Docker
````
```suggestion
### Run FusionAuth Via Docker
```

mark-robustelli commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
````diff
@@ -0,0 +1,283 @@
+mkdir ./public/img && cp ../complete-application/public/img/money.jpg ./public/img/money.jpg
+```
+
+As we will be recreating all of the files in our app directory, please delete all files within `/src/app`.
````
```suggestion
As you will be recreating all of the files in our app directory, please delete all files within `/src/app`.
```

mark-robustelli commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,283 @@
+
+#### Authentication API
+
+Next.js 13.2 introduced <a href="https://nextjs.org/docs/app/building-your-application/routing/router-handlers" target="_blank">Route Handlers</a>, which are the preferred way to handle REST-like requests. In the `Changebank` application we can configure NextAuth.js FusionAuth's provider in a new route handler by creating a file within `src/app/api/auth/[...nextauth]/route.ts`.
```
```suggestion
Next.js 13.2 introduced <a href="https://nextjs.org/docs/app/building-your-application/routing/router-handlers" target="_blank">Route Handlers</a>, which are the preferred way to handle REST-like requests. In the `Changebank` application you can configure NextAuth.js FusionAuth's provider in a new route handler by creating a file within `src/app/api/auth/[...nextauth]/route.ts`.
```

mark-robustelli commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,283 @@
+
+Next.js 13.2 introduced <a href="https://nextjs.org/docs/app/building-your-application/routing/router-handlers" target="_blank">Route Handlers</a>, which are the preferred way to handle REST-like requests. In the `Changebank` application we can configure NextAuth.js FusionAuth's provider in a new route handler by creating a file within `src/app/api/auth/[...nextauth]/route.ts`.
+
+On first load of Next.js this file will make sure that we have all of the correct environment variables. The variables are then exported in an object called `authOptions` which can be imported on the server when we need to get our session using `getServerSession`.
```
```suggestion
On first load of Next.js this file will make sure that you have all of the correct environment variables. The variables are then exported in an object called `authOptions` which can be imported on the server when you need to get our session using `getServerSession`.
```

mark-robustelli commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,283 @@
+#### Login Button
+
+Create a new file in ``src/components/login-button.tsx` that will be used for a button component. Our login button will only be used on the client side so make sure to add `use client` at the top of this file.
+For this button we can use the `signIn` and `signOut` functions from `next-auth/react`. By passing in the session from our pages we can determine if the `Log in` or `Log out` should be shown. (Yeah I know if we could have all decided on Log in or Sign in over the last 20 years that would have been great!)
```
```suggestion
For this button you can use the `signIn` and `signOut` functions from `next-auth/react`. By passing in the session from our pages you can determine if the `Log in` or `Log out` should be shown. (Yeah I know if we could have all decided on Log in or Sign in over the last 20 years that would have been great!)
```

mark-robustelli commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,283 @@
+
+![FusionAuth Login Form](/img/articles/quickstart-javascript-nextjs-web/fusionauth-login.png)
+
+This will then take you back to the application in the API route it will check for a session and appropriately redirect you to the `/account` route when your session has been establised.
```
```suggestion
This will then take you back to the application in the API route it will check for a session and appropriately redirect you to the `/account` route when your session has been established.
```

mark-robustelli commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```suggestion
your application in production, there are some things you're going to want to do.
```

mark-robustelli reviewed (changes requested):
Made a few comments and suggestions, let me know if you have questions.

alex-fusionauth commented:
> Also, need to add the make change page.

Added
https://github.com/FusionAuth/fusionauth-quickstart-javascript-nextjs-web/pull/1

alex-fusionauth commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
````diff
@@ -0,0 +1,283 @@
+```shell
+rm -rf ./src/app && mkdir ./src/app
+```
+
````
I feel like calling out the paths is enough in the files like...
`src/app/api/auth/[...nextauth]/route.ts`

I don't like to prescribe as much if they are on mac vs. windows for commands, or just feel more comfortable in VSCode clicking the make directory button.

alex-fusionauth commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,283 @@
+  lang="ts" />
+
+#### Login Button
+
```
same as above I don't like to prescribe as much if they are on mac vs. windows for commands, or just feel more comfortable in VSCode clicking the make directory button.

alex-fusionauth commented:
@mooreds looks good for final review.

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,277 @@
+</Aside>
+
+FusionAuth will be configured with these settings
+* Your client Id is `c2aeeb1e-65cd-43c5-9326-02dce5e03705`.
```
Please use the same values as the ruby on rails quickstart.

https://fusionauth.io/docs/quickstarts/quickstart-ruby-rails-web

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,277 @@
+
+You can log into the [FusionAuth admin UI](http://localhost:9011/admin) and look around if you want, but with Docker/Kickstart you don’t need to.
+
+### Full Example
```
We don't go through the complete example in any of the other quickstarts.

We should be consistent and either roll this through all of them, or remove it. 

I vote remove, because the complete example process should be outlined in the github repo.

What do you think?

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
````diff
@@ -0,0 +1,277 @@
+npm install
+```
+
+### Walkthrough Example
````
```suggestion
## Authentication
```

mooreds replied:
We want to mimic the ruby on rails structure.

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,277 @@
+
+Accept all of the default prompts like below
+
+![npx installation defaults](/img/articles/quickstart-javascript-nextjs-web/npx-create-next-app.png)
```
Please put in the terminal output here.

mooreds replied:
Also would be awesome if we could specify all of these commands on the command line:

`npx create-next-app@latest changebank --no-tailwind ...`

Is that possible?

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
````diff
@@ -0,0 +1,277 @@
+cd changebank
+```
+
+Install [NextAuth.js](https://next-auth.js.org/getting-started/example), like any tutorial I will show you what this magic does later.
````
```suggestion
Install [NextAuth.js](https://next-auth.js.org/getting-started/example), which simplifies integrating with FusionAuth and creating a secure web application.
```

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,277 @@
+
+#### Styles
+
+Create a new file `src/app/globals.css` copy the below CSS for the changebank application.
```
Not a complete sentence.

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,277 @@
+
+Create a new file `src/app/globals.css` copy the below CSS for the changebank application.
+
+`src/app/globals.css`
```
```suggestion
Here's the contents of `src/app/globals.css`.

```

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,277 @@
+<RemoteCode url={frontmatter.coderoot + "/complete-application/src/app/globals.css"}
+  lang="ts" />
+
+#### Login Button
```
Should review all the header indentation. I think we need to change everything under the walkthrough example to three `#`

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,277 @@
+Create a new file in `src/components/login-button.tsx` that will be used for a button component. Our login button will only be used on the client side so make sure to add `use client` at the top of this file.
+For this button you can use the `signIn` and `signOut` functions from `next-auth/react`. By passing in the session from our pages you can determine if the `Log in` or `Log out` should be shown. (Yeah I know if we could have all decided on Log in or Sign in over the last 20 years that would have been great!)
+
+`src/components/login-button.tsx`
```
Please fix this to be like other file references above.

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,277 @@
+1. Getting the session information - `const session = await getServerSession(authOptions);`
+1. Passing the session to the Login Button - `<LoginButton session={session} />`
+
+`src/app/layout.tsx`
```
File name issue.

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,277 @@
+
+Create a new file `src/app/page.tsx` which will have the Homepage details. Not much here just an image and another Login button.
+
+`src/app/page.tsx`
```
```suggestion
```

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,277 @@
+
+One special note here is that there is a check to see if the session is missing. If it is, you redirect back to the homepage which protects this page on the server for unauthorized access. (You can find the same when a user is logged in on the homepage, it will redirect to `/account`)
+
+`src/app/account/page.tsx`
```
```suggestion
Here's the contents of `src/app/account/page.tsx`.
```

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,277 @@
+### Run the Application
+Run your Next.js application. 
+
+<Aside type="caution">
```
Let's strike this.

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,277 @@
+  lang="ts" />
+
+### Run the Application
+Run your Next.js application. 
```
```suggestion
You can now open up an incognito window and visit the NextJS app at http://localhost:3000/ . Log in with the user account you created when setting up FusionAuth, and you’ll see the email of the user next to a logout button.
```

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,277 @@
+
+Try clicking the `Login` button at the top or center of the screen.
+
+![Changebank Home Page](/img/articles/quickstart-javascript-nextjs-web/homepage.png)
```
We don't have any other images in any of the other quickstarts. Do we want to roll these through? I don't think so, but we can discuss if you feel strongly.

alex-fusionauth replied:
I think this gets back to my visual nature. Adding the quickstart videos solves this for me.

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
Any nextauth docs we can reference here?

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,277 @@
+
+On first load of Next.js this file will make sure that you have all of the correct environment variables. The variables are then exported in an object called `authOptions` which can be imported on the server when you need to get our session using `getServerSession`.
+
+The `FusionAuthProvider` is then provided to `NextAuth` as a provider for any `GET` or `POST` commands that are sent to the `/api/auth/*` route.
```
Since I just removed `src/app` per your instructions, I don't know how to create this file. Do I need to create some directories?

alex-fusionauth replied:
I am not sure I follow, you should just be creating the file `src/app/api/auth/[...nextauth]/route.ts`

Is the issue that you would like to create all these directories separately?

I usually create any file like this...
<img width="369" alt="image" src="https://github.com/FusionAuth/fusionauth-site/assets/139158618/bca78940-4ad9-4240-bf4b-c82672eb3540">

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,277 @@
+For this button you can use the `signIn` and `signOut` functions from `next-auth/react`. By passing in the session from our pages you can determine if the `Log in` or `Log out` should be shown. (Yeah I know if we could have all decided on Log in or Sign in over the last 20 years that would have been great!)
+
+`src/components/login-button.tsx`
+<RemoteCode url={frontmatter.coderoot + "/complete-application/src/components/login-button.tsx"}
```
This was not found.

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,277 @@
+#### Layout
+
+<Aside type="tip">
+If this is your first time using Next.js's App Router, I would highly suggest reading through at least the <a href="https://nextjs.org/docs/app/building-your-application/routing" target="_blank">Routing Fundamentals</a>
```
```suggestion
If this is your first time using Next.js's App Router, I would highly suggest reading through at least the <a href="https://nextjs.org/docs/app/building-your-application/routing" target="_blank">Routing Fundamentals</a>.
```

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,277 @@
+
+#### Account Page (protected)
+
+Create a new file `src/app/account/page.tsx` which will have the Account details.
```
Do I need to create a directory here? I think so.

alex-fusionauth replied:
Is your thought that we should call out all directories and that the relative path is not enough? 🤷‍♂️
<img width="290" alt="image" src="https://github.com/FusionAuth/fusionauth-site/assets/139158618/abe757d6-6ce9-4bb6-a88e-a3ae9a115908">

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,277 @@
+
+![Changebank Home Page](/img/articles/quickstart-javascript-nextjs-web/homepage.png)
+
+
```
Not sure about the rest of this. This feels like an upgrade to our quickstarts but I want to remain consistent. Let's discuss.

alex-fusionauth replied:
That is fair. I think the videos would do more justice :D Happy to chat about it.

alex-fusionauth replied:
removed photos

alex-fusionauth commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,277 @@
+
+Make sure once again that you are in the directory `complete-application/`.
+
+Copy the `.env.example` to `.env.local`
```
done

mooreds commented:
@alex-fusionauth is this ready for a final review?

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,232 @@
+
+
+In this quickstart you are going to build an application with Express and integrate it with FusionAuth.
+You'll be building it for [ChangeBank](https://www.youtube.com/watch?v=pkH-kD73QUM), a global leader in converting dollars into coins.
```
Update the video to https://www.youtube.com/watch?v=CXDxNCzUspM

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,232 @@
+It'll have areas reserved for users who have logged in as well as public facing sections.
+
+The docker compose file and source code for a complete application are available at
+https://github.com/FusionAuth/fusionauth-quickstart-javascript-nextjs-web
```
```suggestion
[https://github.com/FusionAuth/fusionauth-quickstart-javascript-nextjs-web](https://github.com/FusionAuth/fusionauth-quickstart-javascript-nextjs-web).
```

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,232 @@
+
+The `FusionAuthProvider` is then provided to `NextAuth` as a provider for any `GET` or `POST` commands that are sent to the `/api/auth/*` route.
+
+Create a new file named `src/app/api/auth/[...nextauth]/route.ts` and copy the following CSS for the ChangeBank application.
```
CSS?

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,232 @@
+
+### Styles
+
+Create a new file named `src/app/globals.css` copy the below CSS for the ChangeBank application.
```
```suggestion
Create a new file named `src/app/globals.css` and copy the below CSS for the ChangeBank application.
```

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,232 @@
+Create a new file in `src/components/login-button.tsx` that will be used for a button component. Our login button will only be used on the client side so make sure to add `use client` at the top of this file.
+For this button you can use the `signIn` and `signOut` functions from `next-auth/react`. By passing in the session from our pages you can determine if the `Log in` or `Log out` should be shown.
+
+Create a new file named `src/components/login-button.tsx` copy the below code for the ChangeBank application.
```
```suggestion
Copy the below code for the ChangeBank application into  `src/components/login-button.tsx`.
```

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,232 @@
+If this is your first time using Next.js's App Router, you should read through <a href="https://nextjs.org/docs/app/building-your-application/routing" target="_blank">Routing Fundamentals</a>.
+</Aside>
+
+Below you will find the full code for the [Root Layout](https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts#root-layout-required)
```
```suggestion
Below you will find the full code for the [Root Layout](https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts#root-layout-required).
```

I like to copy and paste the entire quickstart into Google docs and run a spell and grammar check on it. That catches stuff like this.

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,232 @@
+
+Below you will find the full code for the [Root Layout](https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts#root-layout-required)
+
+This has the overall structure of our application, the other pages will be added where `{children}` is located.
```
```suggestion
This has the overall structure of our application. The other pages will be added where `{children}` is located.
```

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,232 @@
+
+This has the overall structure of our application, the other pages will be added where `{children}` is located.
+
+There are two items of note:
```
Why are these of note?

alex-fusionauth replied:
now reading it back, maybe the really are not of note...lol

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,232 @@
+1. Getting the session information - `const session = await getServerSession(authOptions);`
+1. Passing the session to the Login Button - `<LoginButton session={session} />`
+
+Create a new file named `src/app/layout.tsx` copy the below code for the ChangeBank application.
```
Not a grammatical sentence.

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,232 @@
+  lang="ts" />
+
+
+### Account Page (protected)
```
```suggestion
### Account Page
```

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```diff
@@ -0,0 +1,232 @@
+
+### Login Button
+
+Create a new file in `src/components/login-button.tsx` that will be used for a button component. Our login button will only be used on the client side so make sure to add `use client` at the top of this file.
```
I think it needs to be `LoginButton.tsx`. That is what it is in `complete-application` and I get an error when working through these steps.

mooreds reviewed (changes requested):
need to add an aside into any blog posts, like we do here: https://fusionauth.io/blog/securely-implement-oauth-in-react (the 'tip' section)

no readme in `complete-application` needed (in https://github.com/FusionAuth/fusionauth-quickstart-javascript-nextjs-web) . I guess it isn't that big a deal if it is generated via npx.

the homepage does not look like the other pages (the section under 'welcome to changebank'). Please consult the ruby on rails web app for what it should look like.

logout url is wrong in kickstart (references port 4200)

can't make change for a non integer value of dollars (4.34 for example). Maybe the input form type?

missing adding the make change page and component from the quickstart instructions (they are present in the complete application)

bradmccarty reviewed (commented):
SEO check is good.

alex-fusionauth commented:
@mooreds 

> need to add an aside into any blog posts,
I am not sure what this is referencing?

> the homepage does not look like the other pages (the section under 'welcome to changebank'). Please consult the ruby on rails web app for what it should look like.
Fixed

This works fine are you on latest?
<img width="658" alt="image" src="https://github.com/FusionAuth/fusionauth-site/assets/139158618/c0d4afdd-fc7e-44b6-b4c2-3f7263c514b8">

> logout url is wrong in kickstart (references port 4200)
Updated

alex-fusionauth commented:
Updated the makechange section

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
```suggestion
* [Hosted pages](/docs/v1/tech/themes/) such as login, registration, email verification, and many more
* [Theming the NextJS Auth pages](https://next-auth.js.org/configuration/pages)
```

mooreds commented on astro/src/content/quickstarts/quickstart-javascript-nextjs-web.mdx:
````diff
@@ -0,0 +1,240 @@
+You can now open up an incognito window and visit the NextJS app at http://localhost:3000/. Log in with the user account you created when setting up FusionAuth, and you’ll see the email of the user next to a logout button.
+
+```shell
+npm run dev
````
I get this error message when running through the instructions.

```
Module not found: Can't resolve '../components/LoginLink'
  3 | import { redirect } from 'next/navigation';
  4 | import { authOptions } from './api/auth/[...nextauth]/route';
> 5 | import LoginLink from '../components/LoginLink';
  6 | 
  7 | export default async function Home() {
  8 |   const session = await getServerSession(authOptions);

https://nextjs.org/docs/messages/module-not-found
```

Looks like we're missing installing a `LoginLink` component. Can you add that step in, please?

```
dan@mooreds fusionauth-quickstart-javascript-nextjs-web % ll complete-application/src/components 
total 24
-rw-r--r--  1 dan  staff   482 Sep 21 13:32 LoginButton.tsx
-rw-r--r--  1 dan  staff   390 Sep 21 13:32 LoginLink.tsx
-rw-r--r--  1 dan  staff  1904 Sep 21 13:32 MakeChangeForm.tsx
dan@mooreds fusionauth-quickstart-javascript-nextjs-web % ll changebank/src/components        
total 16
-rw-r--r--  1 dan  staff   482 Sep 21 13:44 LoginButton.tsx
-rw-r--r--  1 dan  staff  1904 Sep 21 13:46 MakeChangeForm.tsx

```

mooreds commented:
Approved after you fix the issue with the missing component.
